### ISSUE TITLE
$state.current.data is undefined

### ISSUE THREAD

**Issue by adamreisnz** [angular-ui/ui-router#3233]
I need to be able to access the custom data that we set on a state object via `$state.current.data`. However, this property appears to be undefined. Is there any other way to access the data?

If I define a state with a custom property like so:

```
$stateProvider.state('login', {
  auth: true,
});
```

Then I can access `$state.current.auth` and it correctly displays as `true`.
Problem is, that this property does not get inherited by child states.

So resolving to using the `data` property:

```
$stateProvider.state('login', {
  data: {
    auth: true,
  },
});
```

This fixed the inheritance problem, and the property can be accessed in events via for example:

```js
$rootScope.$on('$stateChangeStart', function(event, toState) {
  toState.data.auth === true; //Works, is defined
});
```

But when accessing via the current state it doesn't work:

```js
.run(function($state) {
  $state.current.data.auth === true; //TypeError: Cannot read property 'auth' of undefined 
});
```

Can the data be exposed on the current state as well or accessed somehow?

**Comment by christopherthielen** [comment]
In a `run` block, the current state is the "root" state (`state.name === ""`).  You should use the `$stateChangeStart` (or `onStart()` in ui-router 1.0+) to check if a state needs authentication.

Hope this helps!

**Comment by adamreisnz** [comment]
> In a run block, the current state is the "root" state (state.name === ""). 

But surely the current state changes dynamically, as the state changes in the app, e.g. it's not a static snapshot of the current state, correct?

I apply a listener on the `$rootScope` in a run block, which listens to authentication state change events. If the user becomes *un*authenticated, and the user is currently present in a state which requires authentication, he/she should be returned to the home page.

> You should use the $stateChangeStart (or onStart() in ui-router 1.0+) to check if a state needs authentication.

I am doing that. I don't need it for checking if a state needs authentication before going there, I need to check if the state they are currently on needs authentication, at a time when their authentication status changes (could be a local storage event triggered from another tab).

So in code, something like this (simplified):

```js
.run(function($rootScope, $state) {
  $rootScope.$on('auth.status', function(event, auth) {
    if (!auth.isAuthenticated && $state.current.data.auth === true) {
       $state.go('home');
    }
  });
}
```

**Comment by christopherthielen** [comment]
> But surely the current state changes dynamically, as the state changes in the app, e.g. it's not a static snapshot of the current state, correct? 

Correct, it is updated as the current state changes (it's a mutable global variable).

> Do you recommend another way of obtaining the current state snapshot accurately on such broadcasted events?

Nope... your code looks fine.  What version of ui-router?  Note that not all states will have a `data` property, so you should guard for that. (`$state.current.data && $state.current.data.auth`)

Can you try `$state.$current.data.auth` too?

**Comment by adamreisnz** [comment]
I was using the latest 0.x version I think, e.g. 0.3.2. Yes, I've guarded for the missing data by providing base/default data for my root application route, that seems to work ok.

I've since solved the problem in a different manner; instead of querying $state.current, I save the last state the user successfully navigated to in the auth service, and do the check on that state object instead. That seems to have resolved the issues I was having.
